File: Shared.CQRS/Repository/CommandRepository.cs
using Microsoft.EntityFrameworkCore;
using Shared.Utilities;

namespace Shared.CQRS.Repository;

public abstract class CommandRepository<TEntity, TId, TModel, TDbContext>
    where TEntity : class, IEntity<TId>
    where TId : struct //BaseId<TId>
    where TModel : class
    where TDbContext : DbContext
{
    protected readonly TDbContext DbContext;
    protected readonly IPersistenceMapper<TEntity, TModel> Mapper;

    public CommandRepository(
        TDbContext dbContext, 
        IPersistenceMapper<TEntity, TModel> mapper)
    {
        DbContext = dbContext;
        Mapper = mapper;
    }
    
    public virtual async Task<TModel> AddAsync(TModel model, CancellationToken cancellationToken = default)
    {
        var entity = Mapper.MapToEntity(model);
        return Mapper.MapToModel((await DbContext.Set<TEntity>().AddAsync(entity, cancellationToken)).Entity);
    }

    public virtual async Task<TModel> UpdateAsync(TModel model, CancellationToken cancellationToken = default)
    {
        var entity = Mapper.MapToEntity(model);
        return await Task.FromResult(Mapper.MapToModel(DbContext.Set<TEntity>().Update(entity).Entity));
    }

    public virtual async Task DeleteAsync(TId id, CancellationToken cancellationToken = default)
    {
        var entity = await DbContext.Set<TEntity>().FirstOrDefaultAsync(entity => entity.Id.Equals(id), cancellationToken);
        if (entity is null) throw new InvalidOperationException("Cannot delete entity that does not exist");
        await Task.FromResult(DbContext.Set<TEntity>().Remove(entity));
    }
}

File: IdentityApi/Infrastructure/databases/postgres/repository/base/command_repository.go
package cqrs

import (
	"errors"
	mappers "identity-api/Infrastructure/databases/postgres/mappers/base"
	"log"

	"gorm.io/gorm"
)

type CommandRepository[TEntity any, TId comparable, TModel any] struct {
	dbContext *gorm.DB
	mapper    mappers.PersistenceMapper[TEntity, TModel]
}

func NewCommandRepository[TEntity any, TId comparable, TModel any](dbContext *gorm.DB, mapper mappers.PersistenceMapper[TEntity, TModel]) *CommandRepository[TEntity, TId, TModel] {
	return &CommandRepository[TEntity, TId, TModel]{dbContext: dbContext, mapper: mapper}
}

func (r *CommandRepository[TEntity, TId, TModel]) Add(model TModel) (TModel, error) {
	entity := r.mapper.MapToEntity(model)
	result := r.dbContext.Create(&entity)
	if result.Error != nil {
		log.Println(entity)
		return model, result.Error
	}
	return r.mapper.MapToModel(entity), nil
}

func (r *CommandRepository[TEntity, TId, TModel]) Update(model TModel) (TModel, error) {
	entity := r.mapper.MapToEntity(model)
	result := r.dbContext.Save(&entity)
	if result.Error != nil {
		return model, result.Error
	}
	return r.mapper.MapToModel(entity), nil
}

func (r *CommandRepository[TEntity, TId, TModel]) Delete(id TId) error {
	var entity TEntity
	if err := r.dbContext.First(&entity, "id = ?", id).Error; err != nil {
		return errors.New("cannot delete entity that does not exist")
	}
	return r.dbContext.Delete(&entity).Error
}
